File: core/ad_block.go
package main

import (
	"errors"
	"net"
	"net/netip"
	"net/url"
	"sort"
	"strings"
	"sync"
	"time"

	C "github.com/metacubex/mihomo/constant"
	P "github.com/metacubex/mihomo/constant/provider"
	"github.com/metacubex/mihomo/tunnel"
	"github.com/metacubex/mihomo/tunnel/statistic"
	"golang.org/x/net/idna"
	"golang.org/x/net/publicsuffix"
)

const (
	adBlockEventCapacity          = 500
	adBlockRemoteProviderName     = "__noad_anti_ad"
	adBlockAllowProviderName      = "__noad_allow"
	adBlockLocalBlockProviderName = "__noad_local_block"
	adBlockRuleVersion            = "anti-ad:mihomo.mrs"
)

var noAdRuleProviders = map[string]string{
	adBlockRemoteProviderName:     "anti-ad",
	adBlockLocalBlockProviderName: "manual",
}

type AdBlockEvent struct {
	ID              string    `json:"id"`
	Time            time.Time `json:"time"`
	Host            string    `json:"host"`
	DestinationIP   string    `json:"destinationIp,omitempty"`
	DestinationPort uint16    `json:"destinationPort,omitempty"`
	Network         string    `json:"network"`
	PackageName     string    `json:"packageName,omitempty"`
	UID             uint32    `json:"uid,omitempty"`
	Source          string    `json:"source"`
	Rule            string    `json:"rule"`
	RuleProvider    string    `json:"ruleProvider"`
	RuleVersion     string    `json:"ruleVersion,omitempty"`
}

type AdBlockSnapshot struct {
	Events         []AdBlockEvent `json:"events"`
	TotalBlocked   uint64         `json:"totalBlocked"`
	SessionBlocked uint64         `json:"sessionBlocked"`
	Capacity       int            `json:"capacity"`
	RuleVersion    string         `json:"ruleVersion"`
}

type AdBlockMatchResult struct {
	Matched        bool   `json:"matched"`
	Source         string `json:"source,omitempty"`
	Rule           string `json:"rule,omitempty"`
	NormalizedHost string `json:"normalizedHost,omitempty"`
}

type adBlockEventRing struct {
	mu             sync.Mutex
	events         []AdBlockEvent
	next           int
	totalBlocked   uint64
	sessionBlocked uint64
}

var adBlockEvents = &adBlockEventRing{
	events: make([]AdBlockEvent, 0, adBlockEventCapacity),
}

func (r *adBlockEventRing) add(event AdBlockEvent) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if len(r.events) < adBlockEventCapacity {
		r.events = append(r.events, event)
	} else {
		r.events[r.next] = event
	}
	r.next = (r.next + 1) % adBlockEventCapacity
	r.totalBlocked++
	r.sessionBlocked++
}

func (r *adBlockEventRing) snapshot() AdBlockSnapshot {
	r.mu.Lock()
	defer r.mu.Unlock()

	count := len(r.events)
	events := make([]AdBlockEvent, 0, count)
	for i := 0; i < count; i++ {
		idx := (r.next - 1 - i + adBlockEventCapacity) % adBlockEventCapacity
		if idx >= count {
			continue
		}
		events = append(events, r.events[idx])
	}

	return AdBlockSnapshot{
		Events:         events,
		TotalBlocked:   r.totalBlocked,
		SessionBlocked: r.sessionBlocked,
		Capacity:       adBlockEventCapacity,
		RuleVersion:    adBlockRuleVersion,
	}
}

func (r *adBlockEventRing) clear() {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.events = r.events[:0]
	r.next = 0
	r.totalBlocked = 0
	r.sessionBlocked = 0
}

func recordAdBlockEvent(tracker statistic.Tracker) (*AdBlockEvent, bool) {
	info := tracker.Info()
	if info == nil || !isNoAdRejectedTracker(info) {
		return nil, false
	}

	event := adBlockEventFromTracker(info)
	adBlockEvents.add(event)
	return &event, true
}

func isNoAdRejectedTracker(info *statistic.TrackerInfo) bool {
	if normalizeRuleName(info.Rule) != "ruleset" {
		return false
	}
	if _, ok := noAdRuleProviders[info.RulePayload]; !ok {
		return false
	}
	return chainHasReject(info.Chain)
}

func normalizeRuleName(value string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	value = strings.ReplaceAll(value, "-", "")
	value = strings.ReplaceAll(value, "_", "")
	return value
}

func chainHasReject(chain C.Chain) bool {
	for _, item := range chain {
		if normalizeRuleName(item) == "reject" {
			return true
		}
	}
	return false
}

func adBlockEventFromTracker(info *statistic.TrackerInfo) AdBlockEvent {
	metadata := info.Metadata
	event := AdBlockEvent{
		ID:           info.UUID.String(),
		Time:         info.Start,
		Source:       noAdRuleProviders[info.RulePayload],
		Rule:         info.Rule,
		RuleProvider: info.RulePayload,
		RuleVersion:  adBlockRuleVersion,
	}
	if event.Time.IsZero() {
		event.Time = time.Now()
	}
	if metadata == nil {
		return event
	}

	event.Host = metadata.RuleHost()
	if event.Host == "" {
		event.Host = metadata.String()
	}
	if metadata.DstIP.IsValid() {
		event.DestinationIP = metadata.DstIP.String()
	}
	event.DestinationPort = metadata.DstPort
	event.Network = metadata.NetWork.String()
	event.PackageName = metadata.Process
	event.UID = metadata.Uid
	return event
}

func handleGetAdBlockSnapshot() AdBlockSnapshot {
	return adBlockEvents.snapshot()
}

func handleClearAdBlockEvents() bool {
	adBlockEvents.clear()
	return true
}

func handleNormalizeAdBlockDomain(input string) string {
	domain, err := normalizeAdBlockDomain(input)
	if err != nil {
		return ""
	}
	return domain
}

func handleMatchAdBlockDomain(input string) AdBlockMatchResult {
	domain, err := normalizeAdBlockDomain(input)
	if err != nil {
		return AdBlockMatchResult{}
	}

	providers := tunnel.RuleProviders()
	checks := []struct {
		name   string
		source string
	}{
		{name: adBlockAllowProviderName, source: "allow"},
		{name: adBlockLocalBlockProviderName, source: "manual"},
		{name: adBlockRemoteProviderName, source: "anti-ad"},
	}

	for _, check := range checks {
		provider, ok := providers[check.name]
		if !ok || !providerMatchesDomain(provider, domain) {
			continue
		}
		return AdBlockMatchResult{
			Matched:        true,
			Source:         check.source,
			Rule:           check.name,
			NormalizedHost: domain,
		}
	}

	return AdBlockMatchResult{NormalizedHost: domain}
}

func providerMatchesDomain(provider P.RuleProvider, domain string) bool {
	metadata := &C.Metadata{Host: domain}
	return provider.Match(metadata, C.RuleMatchHelper{})
}

func normalizeAdBlockDomain(input string) (string, error) {
	host := extractDomainHost(input)
	if host == "" {
		return "", errors.New("empty domain")
	}

	host = strings.TrimPrefix(host, "+.")
	host = strings.TrimPrefix(host, "*.")
	host = strings.TrimPrefix(host, ".")
	host = strings.TrimSuffix(host, ".")
	host = strings.ToLower(strings.TrimSpace(host))
	if host == "" || strings.Contains(host, "..") {
		return "", errors.New("invalid domain")
	}
	if _, err := netip.ParseAddr(host); err == nil {
		return "", errors.New("ip address is not a domain")
	}

	ascii, err := idna.Lookup.ToASCII(host)
	if err != nil {
		return "", err
	}
	ascii = strings.ToLower(strings.TrimSuffix(ascii, "."))
	if !hasValidDomainLabels(ascii) {
		return "", errors.New("invalid domain labels")
	}
	if _, err := publicsuffix.EffectiveTLDPlusOne(ascii); err != nil {
		return "", err
	}
	publicSuffix, _ := publicsuffix.PublicSuffix(ascii)
	if publicSuffix == ascii {
		return "", errors.New("domain must not be a public suffix")
	}
	return ascii, nil
}

func extractDomainHost(input string) string {
	value := strings.TrimSpace(input)
	if value == "" {
		return ""
	}
	if strings.Contains(value, "://") {
		parsed, err := url.Parse(value)
		if err == nil {
			return parsed.Hostname()
		}
	}
	if idx := strings.IndexAny(value, "/?#"); idx >= 0 {
		value = value[:idx]
	}
	if host, _, err := net.SplitHostPort(value); err == nil {
		return strings.Trim(host, "[]")
	}
	return strings.Trim(value, "[]")
}

func hasValidDomainLabels(domain string) bool {
	if len(domain) > 253 {
		return false
	}
	labels := strings.Split(domain, ".")
	if len(labels) < 2 {
		return false
	}
	for _, label := range labels {
		if len(label) == 0 || len(label) > 63 {
			return false
		}
		if label[0] == '-' || label[len(label)-1] == '-' {
			return false
		}
	}
	return true
}

func sortedNoAdProviderNames() []string {
	names := make([]string, 0, len(noAdRuleProviders))
	for name := range noAdRuleProviders {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

File: lib/models/ad_block.dart
class AdBlockProps {
  const AdBlockProps({
    this.enabled = true,
    this.allowDomains = const [],
    this.blockDomains = const [],
    this.blockDomainSuffixes = const [],
    this.bypassPackages = const [],
    this.detailedLog = false,
    this.ruleVersion = '',
    this.lastUpdateAt,
  });

  factory AdBlockProps.fromJson(Map<String, Object?>? json) {
    if (json == null) {
      return const AdBlockProps();
    }
    return AdBlockProps(
      enabled: json['enabled'] as bool? ?? true,
      allowDomains: _stringList(json['allowDomains']),
      blockDomains: _stringList(json['blockDomains']),
      blockDomainSuffixes: _stringList(json['blockDomainSuffixes']),
      bypassPackages: _stringList(json['bypassPackages']),
      detailedLog: json['detailedLog'] as bool? ?? false,
      ruleVersion: json['ruleVersion'] as String? ?? '',
      lastUpdateAt: _dateTime(json['lastUpdateAt']),
    );
  }

  final bool enabled;
  final List<String> allowDomains;
  final List<String> blockDomains;
  final List<String> blockDomainSuffixes;
  final List<String> bypassPackages;
  final bool detailedLog;
  final String ruleVersion;
  final DateTime? lastUpdateAt;

  AdBlockProps copyWith({
    bool? enabled,
    List<String>? allowDomains,
    List<String>? blockDomains,
    List<String>? blockDomainSuffixes,
    List<String>? bypassPackages,
    bool? detailedLog,
    String? ruleVersion,
    DateTime? lastUpdateAt,
  }) {
    return AdBlockProps(
      enabled: enabled ?? this.enabled,
      allowDomains: allowDomains ?? this.allowDomains,
      blockDomains: blockDomains ?? this.blockDomains,
      blockDomainSuffixes: blockDomainSuffixes ?? this.blockDomainSuffixes,
      bypassPackages: bypassPackages ?? this.bypassPackages,
      detailedLog: detailedLog ?? this.detailedLog,
      ruleVersion: ruleVersion ?? this.ruleVersion,
      lastUpdateAt: lastUpdateAt ?? this.lastUpdateAt,
    );
  }

  Map<String, Object?> toJson() {
    return {
      'enabled': enabled,
      'allowDomains': allowDomains,
      'blockDomains': blockDomains,
      'blockDomainSuffixes': blockDomainSuffixes,
      'bypassPackages': bypassPackages,
      'detailedLog': detailedLog,
      'ruleVersion': ruleVersion,
      'lastUpdateAt': lastUpdateAt?.toIso8601String(),
    };
  }
}

class AdBlockDomainRule {
  const AdBlockDomainRule({
    required this.value,
    required this.suffix,
  });

  factory AdBlockDomainRule.fromJson(Map<String, Object?> json) {
    return AdBlockDomainRule(
      value: json['value'] as String? ?? '',
      suffix: json['suffix'] as bool? ?? false,
    );
  }

  final String value;
  final bool suffix;

  Map<String, Object?> toJson() {
    return {
      'value': value,
      'suffix': suffix,
    };
  }
}

class AdBlockEvent {
  const AdBlockEvent({
    required this.id,
    required this.time,
    required this.host,
    this.destinationIp,
    this.destinationPort,
    required this.network,
    this.packageName,
    this.uid,
    required this.source,
    required this.rule,
    required this.ruleProvider,
    this.ruleVersion,
  });

  factory AdBlockEvent.fromJson(Map<String, Object?> json) {
    return AdBlockEvent(
      id: json['id'] as String? ?? '',
      time: _dateTime(json['time']) ?? DateTime.fromMillisecondsSinceEpoch(0),
      host: json['host'] as String? ?? '',
      destinationIp: json['destinationIp'] as String?,
      destinationPort: (json['destinationPort'] as num?)?.toInt(),
      network: json['network'] as String? ?? '',
      packageName: json['packageName'] as String?,
      uid: (json['uid'] as num?)?.toInt(),
      source: json['source'] as String? ?? '',
      rule: json['rule'] as String? ?? '',
      ruleProvider: json['ruleProvider'] as String? ?? '',
      ruleVersion: json['ruleVersion'] as String?,
    );
  }

  final String id;
  final DateTime time;
  final String host;
  final String? destinationIp;
  final int? destinationPort;
  final String network;
  final String? packageName;
  final int? uid;
  final String source;
  final String rule;
  final String ruleProvider;
  final String? ruleVersion;

  Map<String, Object?> toJson() {
    return {
      'id': id,
      'time': time.toIso8601String(),
      'host': host,
      'destinationIp': destinationIp,
      'destinationPort': destinationPort,
      'network': network,
      'packageName': packageName,
      'uid': uid,
      'source': source,
      'rule': rule,
      'ruleProvider': ruleProvider,
      'ruleVersion': ruleVersion,
    };
  }
}

class AdBlockSnapshot {
  const AdBlockSnapshot({
    this.events = const [],
    this.totalBlocked = 0,
    this.sessionBlocked = 0,
    this.capacity = 500,
    this.ruleVersion = '',
  });

  factory AdBlockSnapshot.fromJson(Map<String, Object?> json) {
    final events = json['events'] as List? ?? const [];
    return AdBlockSnapshot(
      events: events
          .whereType<Map>()
          .map((event) => AdBlockEvent.fromJson(Map<String, Object?>.from(event)))
          .toList(),
      totalBlocked: (json['totalBlocked'] as num?)?.toInt() ?? 0,
      sessionBlocked: (json['sessionBlocked'] as num?)?.toInt() ?? 0,
      capacity: (json['capacity'] as num?)?.toInt() ?? 500,
      ruleVersion: json['ruleVersion'] as String? ?? '',
    );
  }

  final List<AdBlockEvent> events;
  final int totalBlocked;
  final int sessionBlocked;
  final int capacity;
  final String ruleVersion;

  AdBlockSnapshot copyWith({
    List<AdBlockEvent>? events,
    int? totalBlocked,
    int? sessionBlocked,
    int? capacity,
    String? ruleVersion,
  }) {
    return AdBlockSnapshot(
      events: events ?? this.events,
      totalBlocked: totalBlocked ?? this.totalBlocked,
      sessionBlocked: sessionBlocked ?? this.sessionBlocked,
      capacity: capacity ?? this.capacity,
      ruleVersion: ruleVersion ?? this.ruleVersion,
    );
  }

  Map<String, Object?> toJson() {
    return {
      'events': events.map((event) => event.toJson()).toList(),
      'totalBlocked': totalBlocked,
      'sessionBlocked': sessionBlocked,
      'capacity': capacity,
      'ruleVersion': ruleVersion,
    };
  }
}

class AdBlockMatchResult {
  const AdBlockMatchResult({
    required this.matched,
    this.source,
    this.rule,
    this.normalizedHost,
  });

  factory AdBlockMatchResult.fromJson(Map<String, Object?> json) {
    return AdBlockMatchResult(
      matched: json['matched'] as bool? ?? false,
      source: json['source'] as String?,
      rule: json['rule'] as String?,
      normalizedHost: json['normalizedHost'] as String?,
    );
  }

  final bool matched;
  final String? source;
  final String? rule;
  final String? normalizedHost;

  Map<String, Object?> toJson() {
    return {
      'matched': matched,
      'source': source,
      'rule': rule,
      'normalizedHost': normalizedHost,
    };
  }
}

List<String> _stringList(Object? value) {
  return (value as List? ?? const [])
      .whereType<String>()
      .where((item) => item.trim().isNotEmpty)
      .toList();
}

DateTime? _dateTime(Object? value) {
  if (value is DateTime) {
    return value;
  }
  if (value is String && value.isNotEmpty) {
    return DateTime.tryParse(value);
  }
  return null;
}
